src/EnergyPlus/SurfaceOctree.cc

```diff
@@ -0,0 +1,265 @@
+	bool
+	SurfaceOctreeCube::
+	contains( Surface const & surface ) const
+	{
+		for ( Vertex const & v : surface.Vertex ) { // All surface vertices must be in cube
+			if ( ! contains( v ) ) return false;
+		}
+		return true;
+	}
+
+	// Surfaces Outer Cube Initilization
+	void
+	SurfaceOctreeCube::
+	init( ObjexxFCL::Array1< Surface > & surfaces )
+	{
+		assert( d_ == 0u );
+		assert( n_ == 0u );
+		surfaces_.clear();
+		surfaces_.reserve( surfaces.size() );
+		for( Surface & surface : surfaces ) {
+			if (
+			 ( ! surface.Vertex.empty() ) && // Skip no-vertex "surfaces"
+			 ( ! surface.IsTransparent ) // Skip transparent surfaces
+			) {
+				surfaces_.push_back( &surface );
+			}
+		}
+
+		// No surfaces handler
+		if ( surfaces_.empty() ) {
+			l_ = u_ = c_ = Vertex( 0.0 );
+			w_ = r_ = 0.0;
+			return;
+		}
+
+		// Bounding box corners and center
+		Surface const & surface_0( *surfaces_[ 0 ] );
+		assert( ! surface_0.Vertex.empty() );
+		l_ = u_ = surface_0.Vertex[ 0 ]; // Initialize corners to first vertex of first surface
+		for ( Surface const * surface_p : surfaces_ ) { // Surfaces
+			auto const & vertices( surface_p->Vertex );
+			for ( auto const & vertex : vertices ) { // Expand cube to hold surface vertices
+				l_.min( vertex );
+				u_.max( vertex );
+			}
+		}
+		c_ = cen( l_, u_ ); // Center vertex
+
+		// Expand bounding box to cube with uniform side width
+		Vertex const diagonal( u_ - l_ ); // Diagonal
+		w_ = ObjexxFCL::max( diagonal.x, diagonal.y, diagonal.z );
+		r_ = 0.75 * ( w_ * w_ );
+		Real const h( 0.5 * w_ ); // Half-width
+		l_ = c_ - h;
+		u_ = c_ + h;
+
+		assert( valid() );
+
+		// Branch sub-tree
+		branch();
+	}
+
+	// Valid?
+	bool
+	SurfaceOctreeCube::
+	valid() const
+	{
+		if ( le( l_, c_ ) && le( c_, u_ ) ) {
+			Real const tol2( std::max( std::max( ObjexxFCL::magnitude_squared( l_ ), ObjexxFCL::magnitude_squared( u_ ) ) * ( 4 * std::numeric_limits< Real >::epsilon() ), 2 * std::numeric_limits< Real >::min() ) );
+			if ( ObjexxFCL::distance_squared( c_, cen( l_, u_ ) ) <= tol2 ) {
+				Real const tol( std::max( std::sqrt( std::max( ObjexxFCL::magnitude_squared( l_ ), ObjexxFCL::magnitude_squared( u_ ) ) ) * ( 4 * std::numeric_limits< Real >::epsilon() ), 2 * std::numeric_limits< Real >::min() ) );
+				Vertex const d( u_ - l_ ); // Diagonal
+				return ( std::abs( d.x - w_ ) <= tol ) && ( std::abs( d.x - d.y ) <= tol ) && ( std::abs( d.x - d.z ) <= tol ); // Uniform side widths?
+			}
+		}
+		return false;
+	}
+
+	// Branch to Sub-Tree
+	void
+	SurfaceOctreeCube::
+	branch()
+	{
+		if ( ( surfaces_.size() > maxSurfaces_ ) && ( d_ < maxDepth_ ) ) {
+			// Assign Surfaces to cubes containing them
+			Surfaces surfaces_all;
+			surfaces_all.swap( surfaces_ );
+			for ( auto * surface_p : surfaces_all ) { // Surfaces
+				surfaceBranch( *surface_p );
+			}
+
+			// Compact occupied cube array
+			n_ = 0u;
+			for ( std::uint8_t i = 0; i < 8; ++i ) {
+				SurfaceOctreeCube * & cube = cubes_[ i ];
```

Comment:
@mbadams5 Yes. I use a reference because it changes the `cubes_` entry to `nullptr` after shifting the element down (technically don't need to since we stop at `n_` in all loops but this isn't performance-critical so I'm clearing the shifted entry slots). Here `cube` is just an alias and it doesn't save any compute time: I could just use `cubes_[ i ]` instead of the alias `cube` here if you think that is clearer.

For the second usage in `surfaceBranch` it shouldn't have been a reference and I'll fix that. Thx!